core/domain/suggestion_jobs_one_off.py

```diff
@@ -304,35 +305,63 @@ def map(item):
-                ) = len(values)
-        # Update the suggestion counts.
-        else:
-            if item_category == suggestion_models.SUGGESTION_TYPE_ADD_QUESTION:
-                stats_model.question_suggestion_count = len(values)
-            elif item_category == (
-                    suggestion_models.SUGGESTION_TYPE_TRANSLATE_CONTENT):
-                (
-                    stats_model
-                    .translation_suggestion_counts_by_lang_code[language_code]
-                ) = len(values)
-
-        stats_model.put()
 
-        yield (key, len(values))
+        values_to_yield_after_transaction = []
+
+        def _update_community_contribution_stats_transactional(
+                key, count_value):
+            """Updates the CommunityContributionStatsModel according to the
+            given key and count. The key contains the information for which
+            count to update with the given count value. The model GET and PUT
+            must be done in a transaction to avoid loss of updates that come in
+            rapid succession.
+
+            Args:
+                key: str. A key containing the information regarding which count
+                    to update.
+                count_value: int. The count value for the given key.
+            """
+            item_type, item_category, language_code = key.split(
+                PopulateContributionStatsOneOffJob.KEY_DELIMITER)
+
+            stats_model = (
+                suggestion_models.CommunityContributionStatsModel.get())
+
+            # Update the reviewer counts.
+            if item_type == (
+                    PopulateContributionStatsOneOffJob.ITEM_TYPE_REVIEWER):
+                if item_category == (
+                        PopulateContributionStatsOneOffJob
+                        .ITEM_CATEGORY_QUESTION):
+                    stats_model.question_reviewer_count = count_value
+                elif item_category == (
+                        PopulateContributionStatsOneOffJob
+                        .ITEM_CATEGORY_TRANSLATION):
+                    (
+                        stats_model
+                        .translation_reviewer_counts_by_lang_code[language_code]
+                    ) = count_value
+            # Update the suggestion counts.
+            else:
+                if item_category == (
+                        suggestion_models.SUGGESTION_TYPE_ADD_QUESTION):
+                    stats_model.question_suggestion_count = count_value
+                elif item_category == (
+                        suggestion_models.SUGGESTION_TYPE_TRANSLATE_CONTENT):
+                    (
+                        stats_model
+                        .translation_suggestion_counts_by_lang_code[
+                            language_code]
+                    ) = count_value
+
+            stats_model.put()
+            values_to_yield_after_transaction.append([key, count_value])
```

Comment:
Eh, it seems a bit suspect to me to be modifying values outside of the function during a transaction (especially since that's not rollback friendly, which is an important property of transactions). Could we just return a tuple of the key & count value and then process it at the top-level? ``run_in_transaction`` should return whatever the function under transaction returns.

Reply:
Yes to your explanation, though with two changes: transactions only say "nevermind" in the event of irrecoverable failure (that is, an observed failure for each of the ~3 tries that it attempts), and the rollback only applies to NDB state which is where my earlier concern was: the code would be leaving the transaction with side effects (the variable being changed) even in cases when nothing should have changed (e.g. due to a failure & a rollback).